File: CSES/chessboard-queens.py
# ........
# ........
# ..*.....
# ........
# ........
# .....**.
# ...*....
# ........

ans = 0
cols = [False] * 8
first_diagonal, second_diagonal = [False] * 15, [False] * 15
board = [[x for x in input()] for _ in range(8)]

def solution(row):
    global ans, cols, first_diagonal, second_diagonal, board
    if row == 8:
        ans += 1
        return
    
    for c in range(8):
        if cols[c] or first_diagonal[row + c] or second_diagonal[c - row + 7] or board[row][c] == "*":
            continue
        cols[c] = first_diagonal[row + c] = second_diagonal[c - row + 7] = True
        solution(row + 1)
        cols[c] = first_diagonal[row + c] = second_diagonal[c - row + 7] = False

solution(0)
print(ans)

File: CSES/chessboard-queens.cpp
#include<bits/stdc++.h>
using namespace std;

int ans = 0;
bool cols[8], first_diagonal[15], second_diagonal[15];
char board[8][8];

void res(int row){
    if (row == 8) {
        ans ++;
        return;
    }

    for (int c = 0; c < 8; c++) {
        if (cols[c] || first_diagonal[row + c] || second_diagonal[c - row + 7] || board[row][c] == '*') {
            continue;
        }
        cols[c] = first_diagonal[row + c] = second_diagonal[c - row + 7] = true;
        res(row + 1);
        cols[c] = first_diagonal[row + c] = second_diagonal[c - row + 7] = false;
    }
}

int main(){
    for (int i = 0; i < 8; i++) {
        for (int j = 0; j < 8; j++) {
            cin >> board[i][j];
        }
    }

    res(0);
    cout << ans;

    return 0;
}